token/program-2022/src/extension/confidential_mint_burn/instruction.rs

```diff
@@ -0,0 +1,561 @@
+    ///   6. `[]` The multisig account owner.
+    ///   7.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `MintInstructionData`
+    ConfidentialMint,
+    /// Burn tokens from confidential balance
+    ///
+    /// Fails if the destination account is frozen.
+    /// Fails if the associated mint is extended as `NonTransferable`.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token account.
+    ///   1. `[]` The SPL Token mint. `[writable]` if the mint has a non-zero
+    ///      supply elgamal-pubkey
+    ///   2. `[]` (Optional) Instructions sysvar if at least one of the
+    ///      `zk_elgamal_proof` instructions are included in the same
+    ///      transaction.
+    ///   3. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyCiphertextCommitmentEquality` proof
+    ///   4. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyBatchedGroupedCiphertext3HandlesValidity` proof
+    ///   5. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyBatchedRangeProofU128`
+    ///   6. `[signer]` The single account owner.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` The SPL Token mint. `[writable]` if the mint has a non-zero
+    ///      supply elgamal-pubkey
+    ///   2. `[]` (Optional) Instructions sysvar if at least one of the
+    ///      `zk_elgamal_proof` instructions are included in the same
+    ///      transaction.
+    ///   3. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyCiphertextCommitmentEquality` proof
+    ///   4. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyBatchedGroupedCiphertext3HandlesValidity` proof
+    ///   5. `[]` (Optional) The context state account containing the
+    ///      pre-verified `VerifyBatchedRangeProofU128`
+    ///   6. `[]` The multisig account owner.
+    ///   7.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `BurnInstructionData`
+    ConfidentialBurn,
+}
+
+/// Data expected by `ConfidentialMintBurnInstruction::InitializeMint`
+#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
+#[repr(C)]
+pub struct InitializeMintData {
+    /// Authority used to modify the `ConfidentialMintBurn` mint
+    /// configuration and mint new tokens
+    pub authority: OptionalNonZeroPubkey,
+    /// The ElGamal pubkey used to encrypt the confidential supply
+    pub supply_elgamal_pubkey: OptionalNonZeroElGamalPubkey,
```

Comment:
That's a good point, in my initial proof generation code I just used a dummy key in case the supply elgamal key was none, but the current proof generation logic assumes the supply key to be present for the generation.

Does it make sense to make the confidential supply non optional? Otherwise the token client side logic could also always just generate a random keypair and use that for the proof generation. This would also be fine since the supply related proof contents are only checked / used if a non-none supply elgamal pubkey is present in the mint account. @samkim-crypto wdyt?

Reply:
resolved via other comment: should be non-optional as we want to use the supply to ensure it's within the bound of a u64